[WIP]Allow extracting element-wise loss in softmax

prigoyal opened the pull request:
Summary: Often times, we want to experiment with loss per element (image etc.). This changeset allows getting per element loss as well. This output is optional.

Differential Revision: D15035797

BIT-silence commented on caffe2/operators/softmax_ops.cu:
```diff
@@ -311,6 +311,7 @@ bool SoftmaxWithLossOp<float, CUDAContext>::RunOnDevice() {
 
   auto* avg_loss =
       Output(1, vector<int64_t>(), at::dtype<float>()); // Average loss
+  auto* L = (OutputSize() > 2 ? Output(2) : NULL); // elt_wise loss
```
```suggestion
  auto* L = (OutputSize() > 2 ? Output(2) : nullptr); // elt_wise loss
```

BIT-silence commented on caffe2/operators/softmax_ops.cu:
```diff
@@ -394,7 +395,10 @@ bool SoftmaxWithLossOp<float, CUDAContext>::RunOnDevice() {
         1, scale_ / total_weight, avg_loss_data, avg_loss_data, &context_);
   }
-
+  if (L) {
```
```suggestion
  if (L != nullptr) {
```

xiaomengy reviewed (commented):
Please add related change to CPU op and add some test cases for this change.

prigoyal commented:
thanks @jerryzh168 . There's one test failure but it looks unrelated to this PR. The logs show dlopen related issue. what's the next step to get this PR merged soon? I see: 1) re-run the test 2) land the PR if the test is flaky/expected to fail.

also cc @ezyang in case this test has failed in other PRs too. :)

prigoyal committed:
Allow extracting element-wise loss in softmax (#19579)

Summary:
Often times, we want to experiment with loss per element (image etc.). This changeset allows getting per element loss as well. This output is optional.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/19579

Reviewed By: jerryzh168

Differential Revision: D15035797

Pulled By: jerryzh168

fbshipit-source-id: b4099de94c9b8127eed741032e9a92cbe6a3952c
